x11rb-async/src/rust_connection/mod.rs

```diff
@@ -0,0 +1,1244 @@
+                            return setup_len();
+                        }
+                    };
+
+                    // Wait for the reply.
+                    let cookie = Cookie::<'_, _, EnableReply>::new(self, cookie);
+
+                    let reply = cookie.reply().await.expect("Failed to get reply");
+
+                    // Mark the request as done.
+                    let total = reply
+                        .maximum_request_length
+                        .try_into()
+                        .ok()
+                        .and_then(|x: usize| x.checked_mul(4))
+                        .unwrap_or(std::usize::MAX);
+
+                    *mrl = MaxRequestBytes::Known(total);
+                    total
+                }
+            }
+        })
+    }
+
+    fn parse_error(&self, error: &[u8]) -> Result<x11rb::x11_utils::X11Error, ParseError> {
+        let extensions = future::block_on(self.extensions.read());
+        X11Error::try_parse(error, &*extensions)
+    }
+
+    fn parse_event(&self, event: &[u8]) -> Result<x11rb::protocol::Event, ParseError> {
+        let extensions = future::block_on(self.extensions.read());
+        x11rb::protocol::Event::parse(event, &*extensions)
+    }
+}
+
+impl<S: Stream + Send + Sync> Connection for RustConnection<S> {
+    fn wait_for_raw_event_with_sequence(
+        &self,
+    ) -> Fut<'_, x11rb_protocol::RawEventAndSeqNumber<Self::Buf>, ConnectionError> {
+        Box::pin(self.driven(async move {
+            let get_event = || {
+                self.inner
+                    .lock()
+                    .unwrap()
+                    .conn
+                    .poll_for_event_with_sequence()
+            };
+
+            loop {
+                // Try to get the event.
+                if let Some(event) = get_event() {
+                    return Ok(event);
+                }
+
+                // Register a listener for the reply.
+                let listener = self.new_input.listen();
+
+                // Maybe a packet was delivered while we were registering the listener.
+                if let Some(event) = get_event() {
+                    return Ok(event);
+                }
+
+                // Wait for the next packet.
+                listener.await;
+            }
+        }))
+    }
+
+    fn poll_for_raw_event_with_sequence(
+        &self,
+    ) -> Result<Option<x11rb_protocol::RawEventAndSeqNumber<Self::Buf>>, ConnectionError> {
+        let get_event = || {
+            self.inner
+                .lock()
+                .unwrap()
+                .conn
+                .poll_for_event_with_sequence()
+        };
+
+        // Try to get the event.
+        if let Some(event) = get_event() {
+            return Ok(Some(event));
+        }
+
+        // Maybe we need to be driven. In this case, poll the future once.
+        if let Some(res) = future::block_on(future::poll_once(self.driven(future::pending()))) {
```

Comment:
Huh.... I'm not sure whether this really works. `poll_once` seems quite fragile to me...

This seems like a point for "connect returns a future that does all the reading and then other code becomes simpler".

Reply:
Replaced in the newest code.